glean-core/rlb/src/private/labeled.rs

```diff
@@ -0,0 +1,389 @@
+/// registered in the metrics.yaml file.
+/// Unlike most metric types, `LabeledMetric` does not have its own corresponding
+/// storage, but records metrics for the underlying metric type `T` in the storage
+/// for that type.
+#[derive(Clone)]
+pub struct LabeledMetric<T: AllowLabeled>(
+    pub(crate) Arc<glean_core::metrics::LabeledMetric<T::InnerCoreMetric>>,
+);
+
+impl<T> LabeledMetric<T>
+where
+    T: AllowLabeled,
+{
+    /// The public constructor used by automatically generated metrics.
+    pub fn new(meta: glean_core::CommonMetricData, labels: Option<Vec<String>>) -> Self {
+        let submetric = T::new_inner(meta);
+        let core = glean_core::metrics::LabeledMetric::new(submetric, labels);
+        Self(Arc::new(core))
+    }
+}
+
+#[inherent(pub)]
+impl<T> glean_core::traits::Labeled<T> for LabeledMetric<T>
+where
+    T: AllowLabeled + Clone,
+{
+    /// Gets a specific metric for a given label.
+    ///
+    /// If a set of acceptable labels were specified in the `metrics.yaml` file,
+    /// and the given label is not in the set, it will be recorded under the special `OTHER_LABEL` label.
+    ///
+    /// If a set of acceptable labels was not specified in the `metrics.yaml` file,
+    /// only the first 16 unique labels will be used.
+    /// After that, any additional labels will be recorded under the special `OTHER_LABEL` label.
+    ///
+    /// Labels must be `snake_case` and less than 30 characters.
+    /// If an invalid label is used, the metric will be recorded in the special `OTHER_LABEL` label.
+    fn get(&self, label: &str) -> T {
+        let inner = self.0.get(label);
+        T::from_inner(inner)
+    }
+
+    /// **Exported for test purposes.**
+    ///
+    /// Gets the number of recorded errors for the given metric and error type.
+    ///
+    /// # Arguments
+    ///
+    /// * `error` - The type of error
+    /// * `ping_name` - represents the optional name of the ping to retrieve the
+    ///   metric for. Defaults to the first value in `send_in_pings`.
+    ///
+    /// # Returns
+    ///
+    /// The number of errors reported.
+    fn test_get_num_recorded_errors<'a, S: Into<Option<&'a str>>>(
+        &self,
+        error: ErrorType,
+        ping_name: S,
+    ) -> i32 {
+        dispatcher::block_on_queue();
+
+        crate::with_glean_mut(|glean| {
+            glean_core::test_get_num_recorded_errors(
+                &glean,
+                self.0.get_submetric().meta(),
+                error,
+                ping_name.into(),
+            )
+            .unwrap_or(0)
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::ErrorType;
+    use crate::common_test::{lock_test, new_glean};
+    use crate::destroy_glean;
+    use crate::private::{BooleanMetric, CounterMetric, LabeledMetric, StringMetric};
+    use crate::CommonMetricData;
+
+    #[test]
+    fn test_labeled_counter_type() {
+        let _lock = lock_test();
+
+        let _ = new_glean(None, true);
```

Comment:
This drops the returned `TempDir` immediately, thus removing it. This kinda works here, because we keep the database open, but really we should keep the directory around for this test.

```suggestion
        let _t = new_glean(None, true);
```

Reply:
same in the following tests.